### DISCUSSION TITLE
chessboardHybridStereo: using pre-calibrated camera data to improve distortion coefficients

### DISCUSSION THREAD

**Discussion by KevinCain** [decadenza/SimpleStereo#30] [General]
We have the unusual case where one camera in the stereo rig provides its own undistortion based on its own manufacturer calibration. Like other vision libraries, OpenCV’s ‘stereoCalibrate’ is meant to take unprocessed photos as input, preferably with the same camera specs. Even if we give ‘stereoCalibrate’ coefficients with zero values for one or both cameras, it will still merrily alter both sets of coefficients as it tries to find the best extrinsics estimate. We can lock the intrinsics with the ‘cv2.CALIB_FIX_INTRINSIC’ flag, however this locks _both_ cameras.

We are therefore calibrating our left camera in two passes using [chessboardHybridStereo](https://github.com/KevinCain/SimpleStereo/blob/6c5235116263caf6e8ebccf3534ddaf5f1ed3518/simplestereo/calibration.py#L102), while setting the right camera distortion coefficients to zero:
**Pass One:** We use the distortion coefficients estimated from a set of 140 images for the left camera: these photos capture a wide range of distortion across the field of view and have a mean reprojection error of 0.037811.
**Pass Two:** We then use the intrinsic parameters of the left camera from a separate 21-images set with mean reprojection error of 0.017860. For this image set, both left and right cameras can see the chessboard. The right camera's distortion coefficients we set as zeros, assuming that any physical lens distortion has been corrected by the camera's internal processing, which as above is a ‘black box’ for us since we do not have code for its methods.

Our results aren't good (see below) -- is this approach valid? Notes follow:

Because we lock intrinsics during stereo calibration, we also perform subpixel refinement on the corner points and apply distortion correction for these chessboard corner points which would normally be freely optimized in ‘stereoCalibrate’. Also, because both cameras are fixed we must also estimate the right camera pose from the above set of 21 images, to pass its matrix to ‘cv.stereoCalibrate’.

To complicate things, the placement of cameras ensures there is little overlap between camera frustums as described in
[this forum post](https://github.com/decadenza/SimpleStereo/issues/27). The result is that we cannot shoot a group of calibration images where the chessboard occupies most parts of the frame, since that will cause the chessboard to exit frame in the other photo in the stereo pair.

My steps follow, with code available in [this branch of my SS fork](https://github.com/KevinCain/SimpleStereo/tree/HybridStereo).

**‘calib.py’** - Chessboard calibration and reprojection for the two sets of images

**‘buildStereoRig.py’** - Call 'chessboardHybridStereo' for stereo calibration (yields ~2.4 pixels net error), and two reprojection tests, one using computed matrices and another calling PnP to estimate camera pose for reprojection. The distortion coefficients are indeed different between our two sets of images:

> Distortion Coefficients Left Camera 21 Images: [[-0.24081981  0.05132648 -0.00028436  0.00071607 -0.00523626]]
> Distortion Coefficients Left Camera 140 Images: [[-0.23439552  0.04510877 -0.00033863  0.00042451 -0.00333957]]
> Distortion Coefficients Right Camera 21 Images: [[ 1.49431227e-01 -6.03047996e-01  4.43286004e-04  2.49145242e-03
>    7.55302141e-01]]

**‘display_images.py’** - Display epipolar lines on corrected images -- note they are close but not close enough for matching: 
![image](https://github.com/decadenza/SimpleStereo/assets/5424003/f3f172c5-4b70-4928-8edc-0a198561b28c)

**‘rectifyRig.py’**

**‘imageRectify.py’**

**Comment by KevinCain** [comment]
Here is a [temporary link to the image files used in the python code above](https://www.transfernow.net/dl/20231111UlKFtZCz).

/left: (140) photos [used to calculate distortion coefficients for left camera]
/intrinsics_sync/left: (23) photos [used to estimate extrinsics between left and center cameras]
/intrinsics_sync/center: (23) photos [used to estimate extrinsics between left and center cameras]

**Comment by decadenza** [comment]
> We have the unusual case where one camera in the stereo rig provides its own undistortion based on its own manufacturer calibration. Like other vision libraries, OpenCV’s ‘stereoCalibrate’ is meant to take unprocessed photos as input, preferably with the same camera specs.

The first workaround I can guess, is to capture the images from the first already-calibrated camera (let's call it A), undistort these images according to manufacturer specs and then use them in conjunction with camera B with common settings.

If the manufacturer correction is almost perfect, resulting distortion coefficients for camera A will be almost zero. The only disadvantage is longer optimisation time for the OpenCV algorithm but this shouldn't be an issue.

> We are therefore calibrating our left camera in two passes using [chessboardHybridStereo](https://github.com/KevinCain/SimpleStereo/blob/6c5235116263caf6e8ebccf3534ddaf5f1ed3518/simplestereo/calibration.py#L102)

I looked at the code and I see you are loading intrinsics for both cameras. Are both of them already calibrated or only one?
If only camera A is calibrated, you can use the above workaround. If both of them are calibrated, you can either ignore the manufacturer calibration values and re-calibrate or use them and fix the intrinsics to zero.
Could you please explain this?

> To complicate things, the placement of cameras ensures there is little overlap between camera frustums as described in
https://github.com/decadenza/SimpleStereo/issues/27.

This issue is pure geometrically. Only minor obvious things can be done: move the cameras, change focal length, change chessboard size, etc... 

Converting this thread to a discussion as it is not related to SS issue.

**Reply by KevinCain** [reply]
Thanks @decadenza for your notes and questions.

First, to answer you clearly:  I'm computing intrinsics for both cameras (our self-processing RGB camera you note as 'A' and a second monochrome camera 'B' which gives distorted output) since I use `CALIB_FIX_INTRINSIC` to accommodate cam 'A'. As above `cv2.stereoCalibrate` seems to have no provision to lock only one camera's intrinsics: therefore it seems we need a good estimate of intrinsics for both cameras in order to be able to solve for the extrinsics between cameras in the rig.

I should have emphasized that our self-processing camera 'A' doesn't provide specs or metadata; it provides photos that it has internally undistorted and presumably cropped using its own methods, leaving no log to recover the original distorted frame aspect, sensor size, or intrinsics. Here's a frame from that camera -- note that straight lines are already fairly straight:
![rgb (82)](https://github.com/decadenza/SimpleStereo/assets/5424003/ed86f9f2-31cc-4379-bd54-3a8834442f0d)

If I don't lock intrinsics before building the stereo rig calibration in SS, `cv2.stereoCalibrate` adjusts the radial distortion coefficients yielding this pincushion warp (cam 'B' left, cam 'A' right): 
![image](https://github.com/decadenza/SimpleStereo/assets/5424003/5fa3cd12-fd14-4af2-89b5-e9012177b1d5)

Using SS to calibrate camera 'A' by itself, openCV takes a long while to converge and gives these results:
> Camera matrix:  [[3.21889753e+03 0.00000000e+00 2.06792457e+03]
>  [0.00000000e+00 3.21526427e+03 1.55479667e+03]
>  [0.00000000e+00 0.00000000e+00 1.00000000e+00]]
> Distortion coefficients:  [[ 1.49426125e-01 -6.03025492e-01  4.43068394e-04  2.49410179e-03
>    7.55304675e-01]]

Given the geometry of this stereo rig, we have fairly little overlap between cameras and therefore shooting photos where we have an in-focus chessboard in both frames is constrained to a fairly small part of each respective frame.
